Fix/1443 remove the JS runtime from threshold calculations

oleiade opened the pull request:
<!--


  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/grafana/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->
This Pull Request addresses the minimum scope described in #1443 and replaces the current thresholds' expression evaluation implementation. It drops the reliance on the [Goja JS runtime](https://pkg.go.dev/github.com/dop251/goja#Runtime), in favor of a pure-go based implementation. With this PR, thresholds are parsed, and evaluated directly in Go, and all JS runtime-related threshold code has been dropped. It sticks to the existing format, and does not introduce any of the design and format changes proposed in #1443 and #1441.

Design notes regarding the PR's content:
  - The introduction of a minimalistic parser combinators library, taken and adapted from [Jeffhail benthos' bloblang implementation](https://github.com/Jeffail/benthos/). The design choice of combinators was motivated by their ease of use, of extension, and general maintainability and performance. Looking at #1443 and #1441 it also appeared like the best choice when it comes to further improvements we might want to make. The choice to extract the code from Bloblang was motivated by the fact that this was definitely the cleanest and idiomatic implementation I could find; but it was unfortunately (to my knowledge) not offered as a dedicated  library module. I also brought a few improvements in the process, by modifying and extending the existing API to fit K6's use-case.
  - I put the library's code in `pkg`. I was not entirely sure what the best place would be for code that's generic enough to be a good independent library, but not big enough to make it into a dedicated repo. `pkg` is often used as the place where projects put their public reusable code. I'm not hung onto it at all. Please propose an alternative if you prefer to put it in another place 👀 
  - I added a sink map to the `Thresholds` struct, and pass it around to children `Threshold` instances when they need to be evaluated. It's the simplest thing I could think of, but with my limited understanding of the current design of K6, I might have missed another way of doing it.

Let me know what you think 👀 🐻

oleiade commented:
I'd be really interested in your thoughts on how I approached testing @inancgumus; I'm all ears for any recommendations on that topic specifically 👂🏻 🙏🏻

MStoykov commented:
I haven't gone through the code, yet, but I would expect expressions like `1+1=2` and so on to possibly be valid. But also from UX perspective, it will be better if we require that something from the metric is used and if not to error that the threshold isn't based on the metric. 

Given that, for the purposes of the tests, you can rewrite (as you seem to have done) in a way that you get the test to pass -where it requires that the threshold fails just to write a threshold that will fail when evaluated but still based on the metric.

yorugac commented:
@oleiade I've only started digging into this, hopefully will finish tomorrow :sweat_smile: But I have a preliminary question: the work you reference in the PR description (Jeffhail benthos' bloblang) -- should it be referenced in the code as well? I imagine it depends on how much you adapted it...

Also, CI failure shows:
```
--- FAIL: TestClient (0.01s)
    --- FAIL: TestClient/ReflectUnregistered (0.03s)
        client_test.go:650: 
            	Error Trace:	client_test.go:650
            	            				client_test.go:687
            	Error:      	"can't list services: can't send request: EOF at go.k6.io/k6/js/common.Bind.func1 (native)" does not contain "rpc error: code = Unimplemented desc = unknown service grpc.reflection.v1alpha.ServerReflection"
            	Test:       	TestClient/ReflectUnregistered
FAIL
```
This doesn't look like a flaky one... Did it happen before? I can't repeat it on the current master :confused:

oleiade commented:
@yorugac Sorry for the huge chunk of code I've just dropped 🙏🏻 I prefer to make smaller iterative PRs in general, but I found it hard to avoid in this case. You're right about mentioning Jeff Hail's work somehow in the code. I wanted to find out what would be the best way to do that, and it somehow slipped out of my mind. I've adapted and extended the API, and modified some implementation here and there, but a big part of the code is still verbatim. I would like to add a copy of his copyright license at the top of the file (as is the MIT license etiquette, I think?); and mention that the base code comes from bloblang's implementation. Does that sound like a good approach.

Regarding the failing test, as I haven't touched the code since yesterday, and only rebased my PR on master, I can only assume that either the failing test comes from master, or that it impacted my code's behavior somehow. I'll look into it, thanks for pointing it out! 👍🏻

oleiade commented:
@yorugac I've modified the `combinators` package description to mention that most of the code comes from jeffhail's work. I also added their License to the top of the file, as the MIT requests. Do you find it sufficient?

yorugac reviewed (changes requested):
Thank you for the clarification on licenses, @oleiade ! It is now clear enough to me what happens with the code adaptation in this PR :+1: But I'm not sure about how it _should_ be done in k6 code base. I.e. should there be 2 full license headers as is currently done in `combinators.go`: it takes a lot of space and IIRC, there was some talk about dropping all of those headers? The only precedent in k6 that I know of ATM is the tc39 files and those are left without a header.
@na--, @MStoykov, could you please add your opinion on what should be done in this case with license / authorship?

On the review part. `combinators` pkg looks quite nice and `Result` is reminding me of some other languages :smile: I'm not sure about its name though: maybe rename it to just `parser`? I currently assume "combinators" came from the original code but it doesn't make much sense in k6 context IMHO. I don't have a strong opinion about `pkg` folder except there is only `lib` at the moment: are we moving `lib` -> `pkg`? It doesn't make sense to me to have 2 folders without understanding the difference between them.

Overall, I like how this change shapes up a lot! Theo, please see the comments: there are some suggestions about renamings and some requests about logic and clarifications. And one more thing I've been wondering about: should some benchmarking be done here?

yorugac commented on stats/thresholds.go:
```diff
@@ -122,9 +213,9 @@ func (tc thresholdConfig) MarshalJSON() ([]byte, error) {
-	Runtime    *goja.Runtime
 	Thresholds []*Threshold
 	Abort      bool
+	Sinked     map[string]float64
```
One more thing: since `Sinked` is not used anywhere else, I suggest making it `sinked`.

oleiade commented:
@yorugac Thanks for the review! I'll look into your comments and address them when I can. 

Regarding the naming, I don't think `parser` would make a lot of sense considering that parser combinators are actually a "generic" parsing technique/toolbox you build parsers on top of: as opposed to a recursive descent or LRLA parser for instance, which you tailor specifically to the format you're trying to parse. I also kept it separate in the spirit of a) being able to reuse it to extend the threshold's parsing in the future (to fit with the initial proposal this PR partly answers to) and b) making it possible for a potential third-party repo to use it directly from our repo by adding `github.com/grafana/k6/pkg/combinators` to their dependencies. Initially I considered naming it `crocombine` and putting in a separate repo...

Regarding benchmarking, I've thought about it, without never actually getting to implementing it. Now that you ask about it, I'll add some :)

oleiade commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,230 @@
+import (
+	"fmt"
+
+	c "go.k6.io/k6/pkg/combinators"
```
I agree with you, this is a left over. Explicit is better than implicit 👍🏻

oleiade commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,230 @@
+		}
+
+		parsed, ok := res.Payload.([]interface{})
+		if !ok {
+			err := fmt.Errorf("failed parsing percentile expression; " +
+				"reason: converting ParsePercentile() parser result's payload to []interface{} failed",
+			)
+			res.Err = c.NewFatalError(input, err, "percentile")
+		}
```
That's an excellent idea, I'll do that 💯

oleiade commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,230 @@
+			)
+			res.Err = c.NewFatalError(input, err, "percentile")
+		}
+
+		percentile := fmt.Sprintf("%s%g%s", parsed[0].(string), parsed[1].(float64), parsed[2].(string))
```
That's a very good point, I do not check that the percentile is with bounds `[0, 100]` at the moment. Going to add that + a test for it.

On the front of precision, the `%g` formatter in the `Sprintf` call ensures to limit the decimal precision to the minimum necessary. For instance, `99.0` would be represented as `99`, but `99.1` would remain `99.1` while `99.10` would be truncated to `99.1`.

yorugac replied:
Ah, I'm sorry, I should have been more verbose in this comment about precision :sweat_smile: I meant not the precision in this particular line but in all steps of thresholds processing together: string -> float64 -> string. I.e. it's worth to consider if there can be any loss / misrepresentation at some point and if it can, IMO, it should be either fixed or documented. I can't say that I see the problem with conversions right now but I'd have run some additional tests for edge cases, etc., to be sure. Just a suggestion :slightly_smiling_face: 

However, this does work with current `master`:
```
http_req_failed: ['rate<123e5'],
http_req_duration: ['p(34)<123e-5'],
```
But I don't think this would work in this PR because exponent is not processed in the value. To be clear, I'm not saying this _must_ work... e.g. was anyone actually using this form of floats in their scripts? Maybe it can be dropped? But this is probably something to decide / investigate more.

oleiade commented on stats/thresholds.go:
```diff
@@ -26,70 +25,162 @@ import (
+	var ok bool
+	parsed, ok := result.Payload.([]interface{})
+	if !ok || len(parsed) != 3 {
+		return nil, fmt.Errorf("parsing threshold condition %s failed; reason: malformed expression", expression)
```
You're right! I'll make it more explicit 👍🏻

oleiade commented on stats/thresholds.go:
```diff
@@ -26,70 +25,162 @@ import (
+	}
+
+	// Unpack the various components of the parsed threshold expression
+	method, ok := parsed[0].(string)
+	if !ok {
+		return nil, fmt.Errorf("the threshold expression parser failed; " +
+			"reason: unable to cast parsed aggregation method to string",
+		)
+	}
+	operator, ok := parsed[1].(string)
+	if !ok {
+		return nil, fmt.Errorf("the threshold expression parser failed; " +
+			"reason: unable to cast parsed operator to string",
+		)
+	}
+
+	value, ok := parsed[2].(float64)
+	if !ok {
+		return nil, fmt.Errorf("the threshold expression parser failed; " +
+			"reason: unable to cast parsed value to underlying type (float64)",
+		)
+	}
```
In this case, the threshold and parser combinators libraries should protect us from this. These checks are really here for protecting our future selves, if we were to make some breaking changes in either the threshold parser or combinators code, and to make the linter agree with the casts.

oleiade commented on stats/thresholds.go:
```diff
@@ -26,70 +25,162 @@ import (
+		passes = lhs < t.parsed.Value
+	case "==":
+		passes = lhs == t.parsed.Value
+	case "===":
+		// Considering a sink always maps to float64 values,
+		// strictly equal is equivalent to loosely equal
```
In this PR I try to fit as close as possible to the existing implementation, without changing the threshold expression format. As they were written in JS before, `===` is valid there, although it makes no sense in Go, which has a static type system. So supporting `===` is really there for compatibility reason, but as we a predefined type under the hood `==` and `===` are therefore equivalent.

yorugac replied:
Got it, that's clear, thanks! But now I wonder if some people might expect a different behavior when using `==` and `===` in thresholds. I.e. isn't it more straight-forward to have just 1 operator without any ambiguity? :thinking:
I think this point could use input (or clarifications if it was already discussed in the past) from others :slightly_smiling_face:

MStoykov replied:
I think it's perfectly fine for `===` to be equal ( :stuck_out_tongue_winking_eye: ) to `==` and it's important to keep this as compatible with what is in the cloud and this is supported there.

MStoykov replied:
also I would prefer for this and the previous case to be merged ;)

oleiade commented on pkg/combinators/combinators_test.go:
```diff
@@ -0,0 +1,568 @@
+* You should have received a copy of the GNU Affero General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+ */
+package combinators
```
I think you're right, it's probably fine to have it only once where the package description lives 👍🏻

oleiade commented on pkg/combinators/error.go:
```diff
@@ -0,0 +1,88 @@
+* You should have received a copy of the GNU Affero General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+ */
+
+package combinators
```
I think you're right, it's probably fine to have it only once where the package description lives 👍🏻

oleiade commented on stats/thresholds.go:
```diff
@@ -26,70 +25,162 @@ import (
+// operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
+// float               -> digit+ (. digit+)?
+// digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
+// whitespace          -> space | tab
```
That's a good question. It's probably one of the blind spots of the BNF. The parser as I wrote it discards `\n`, `\r`, `\t`, and `\r\n`. I added really as sentinels, to be sure that even if programmers were to add them to their expression, they wouldn't impact the execution. The goal of a BNF being to be as explicit as possible on the format, I would actually add a definition for newline, as being `newline -> "\n" | "\r\n"`, and add it to assertion as optional. Does that address your question? 🐻

oleiade replied:
Here's an updated BNF to explicit what I meant:

```
assertion           -> aggregation_method whitespace* operator whitespace* float newline*
aggregation_method  -> trend | rate | gauge | counter
counter             -> "count" | "sum" | "rate"
gauge               -> "last" | "min" | "max" | "value"
rate                -> "rate"
trend               -> "min" | "mean" | "avg" | "max" | percentile
percentile          -> "p(" float ")"
operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
float               -> digit+ (. digit+)?
digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
whitespace          -> space | tab
tab                 -> "\t"
space               -> " "
newline		 -> linefeed | crlf
crlf 		-> carriage_return linefeed
linefeed		-> "\n"
carriage_return	-> "\r"
```

yorugac replied:
Yes, I missed that yesterday: currently tabs are allowed inside threshold expression and newline is allowed in the end as you described.
So this is again more a question of design: _should_ there be tabs or newlines there? What's the use case that someone would need those chars in threshold expression? I was more thinking in line of simplifying it as much as possible and maybe help people avoid some mistakes in writing tests. I guess it can be seen both ways :thinking: 

And I agree that BNF is a very nice way to be clear on expectations in this case :+1:

oleiade replied:
That's a good question. A decision needs to be made here. As with the rest of this PR, my approach has been to get as close as possible to what is documented and already implemented (with my probably limited view of it still). I would tend to think that having the parser discard spaces and tabs is a nice to have, as it makes it permissive to "not perfect" input. I have no opinion on the newlines. Furthermore, I don't think I have the experience/knowledge of the test scripts and how they're used in the wild yet to be an able to decide, so I'll let you folks decide :)

MStoykov replied:
if the cloud supports tabs we should support them as well, if not - not

oleiade replied:
Looking at the cloud's parser, they drop spaces, but do not seem to touch at tabs, nor newlines actually: I assume they produce an error in that case.

See https://github.com/grafana/k6-cloud-lib/blob/develop/k6_lib/utils/thresholds.py#L66

Thus I will make the parser *not* ignore tabs nor new lines.

yorugac commented on stats/thresholds.go:
```diff
@@ -26,70 +25,185 @@ import (
+	// expression
+	lhs, ok := sinks[t.parsed.AggregationMethod]
+	if !ok {
+		return false, fmt.Errorf("unable to apply threshold %s over metrics; reason: "+
+			"no metric supporting the %s aggregation method found",
+			t.parsed.AggregationMethod,
+			t.parsed.AggregationMethod)
```
```suggestion
		return false, fmt.Errorf("unable to apply threshold %s over metrics; reason: "+
			"no metric supporting the %s aggregation method found",
			t.Source,
			t.parsed.AggregationMethod)
```

yorugac commented on stats/thresholds.go:
```diff
@@ -138,60 +252,54 @@ func NewThresholds(sources []string) (Thresholds, error) {
+func (ts *Thresholds) Run(sink Sink, duration time.Duration) (bool, error) {
+	// Update the sinks store
+	ts.sinked = make(map[string]float64)
+	f := sink.Format(duration)
```
Looking at this with fresh eyes, I don't think custom percentiles are supported right now, are they? So here, `Format` in case of `Trend` will only calculate `P(90)` and `P(95)` by default. But what if user specifies something else? Like I just wrote in another comment:
```
http_req_duration: ['p(34)<123e-5']
````
It seems to me the old version was doing this in JS, with `jsEnvSrc` substitute...

Fixing this would likely require a cast to `TrendSink` as `P()` is not a part of `Sink` interface :slightly_frowning_face: I had to cast this in PRW extension too.

oleiade replied:
That's a good question! I must admit, I focused on what ended up in the sinks, and not so much on how it ended in there. I'm going to run a bunch of tests, see what happens, and I'll let you know. Not only that, but I might ask you for help if things go south 🙏🏻

MStoykov replied:
yes it needs to support `p(<random float between 0 and 100>)`, and yes previously that was done through the "reexporting" of the Trend's [P method](https://github.com/grafana/k6/blob/14079775b9bad0b8b2f503a7fa1e787fbc5cba36/stats/sink.go#L113-L131)

oleiade replied:
With my recent changes, the parser supports and verifies that supplied values are floats in the range `[0, 100]` in either integer or decimal notation. It doesn't support the *exponent* notation of floating numbers, though. Should it?

MStoykov replied:
It seems to me that this code still doesn't support percentiles?!?

I would also, looking at this code and the other places where `sink.Format` is used, I kind of want to remove this method. It's in most cases used as a way to get info about values for the given sink, but then we also cast the sink to its concrete type in order to get even more info from it. 

Given that there are 4 sink types, I feel like a switch in all cases (as the fields might not be the same) might be better :thinking: .

Especially as it turns out that the [CounterSink](https://github.com/grafana/k6/blob/bd221f75725dbf7b62590829eea9f49029d1508a/stats/sink.go#L61) does actually return a `rate` but [the summary code overwrites it](https://github.com/grafana/k6/blob/bd221f75725dbf7b62590829eea9f49029d1508a/js/summary.go#L57-L61) possibly with the same value :facepalm:  

And in the REST API where it is used "correctly" we have a [minor bug](https://github.com/grafana/k6/issues/1253) about Trends and also doesn't show the same stuff as in the summary :facepalm: .

Fixing all of this out of the scope for this PR, but I kind of feel it might be better to *not* depend on what `Format` returns in the thresholds

oleiade replied:
I have opened a dedicated issue: #2320. Please don't hesitate to add information you find missing @MStoykov. Resolving this discussion; but let me know if I've missed something that still needs addressing 🙏🏻

MStoykov commented:
Having had (not so) fast of a look through the changes it looks pretty nice, but I kind of feel like the whole combinators based parser is a bit of overkill:
- while there are some ideas of extending the thresholds, I very much doubt this will happen soon, or that it will use the same syntax. Any change to this (or any future) threshold definition syntax needs to also be done in the k6 cloud and that in general means that we will likely need quite a big redesign. My expectation is that we will completely redesign the system after the metric system redesign and likely whatever is currently being done will *only* be kept for backwards compatibility
- reusing someone else's code like this will likely require it to not be licensed as theirs and not have a copyright by us. This I guess is more of a legal problem but does also (at least for me) tip the balance to not using it if possible.
- I wrote [this](https://go.dev/play/p/OxDzfJnYMuE) in 15 minutes, in general, because of some of the other comments here. IMO it is some error handling away from being production-ready(so the remaining 80% ;) ). Also happens to be one screen long and at least to me *a lot* more understandable than the 1000+ lines of code and tests for the parser you wrote.  The main place that I would expect your approach will win (apart from extensibility, although even that is dependant on the change) is error handling - I would expect your parser will give nicer errors, but even that can be .. improved. And given the simplicity of it all I really don't see why we would want 1000 lines + of parser :( 

I will go through the rest of the comments and comment on those, but just wanted to start the discussion on that as this will probably cut in half the amount of code added by this PR 

cc @na--

oleiade commented:
Hey @MStoykov thanks for your input 🙏🏻 

- Regarding the scope of the changes. I started implementing this, trying to follow what's described in #1443 and #1441 as close as possible. Hence, the initial choice to start with a more fully-featured parsing approach. It's only later that it was clarified to me that we didn't actually intend to implement the scope described in these issues (?), and at that point I already had used the combinators.
- Regarding the licensing, I followed what the MIT License requires: "The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.". But indeed, I don't know what repercussions this leads to when it comes to the compatibility with our own license. PS: I could also republish the code in my own name if relevant.
- Regarding the usage of combinators and of a dedicated parsing library all in all, and if we keep in mind that I wrote that initially with a larger scope in mind, you're right, a smaller parser could be written, and I'm open to do that. But I'd like to point out some upsides of the approach I took:
- its flexibility (you can build any kind of parsers using that technique, without having to rewrite a whole new parser every time, it's really more of a framework than anything else).
- If you put the combinators toolkit aside, the parsing code (threshold_parser) is actually quite small.
- error management; which I think is critical in terms of UX (this parser tells you which part of the expression is wrong, why, and at what character).

All in all, I must say I'm still a bit confused by the scope and the plan regarding thresholds, as I've heard "different bell sounds" 🇫🇷. So I'll trust you folks on what's the best approach here:
- Would you prefer I drop the combinators approach and replace it with a custom smaller parser indeed?
- Does the BNF I defined works? I've aligned with what's in our documentation, but If relevant, who should I talk to, to align it with the cloud's?

na-- commented:
_(Disclaimer: I haven't reviewed the whole PR yet, just skimmed it... my current answer is mostly about the latest discussion :arrow_up: about the custom parser)_

TLDR version: I generally agree with @MStoykov.

Here's my rationalle... We don't know how the k6 thresholds of the future will look like... [#1441](https://github.com/grafana/k6/issues/1441) is one suggestion for one aspect of them, there are various improvement ideas in [#1136](https://github.com/grafana/k6/issues/1136) and some of its linked issues and in [this comment](https://github.com/grafana/k6/issues/1443#issuecomment-626957527) (probably others as well), [#1831](https://github.com/grafana/k6/issues/1831) will almost certainly also affect how we do things...

So, we don't know _how_ thresholds 2.0 will look like, but it's almost certainly going to be different than the current thresholds. Our current syntax is just too limited to support a lot of the things we want. Though, even after we have better thresholds and deprecate the current syntax, we probably won't ever stop supporting it... It's such a major feature that all old k6 scripts should continue to work. We'll just silently (or with a warning) support the old format, internally translating it to the new feature (since its capabilities will be a superset of the current ones), but we probably won't ever develop it or add new features to it.

And that's the crucial reason why I think we should go with simplicity here. There's no need to develop a fancy custom parser for something that is likely going to be deprecated soon. The current format should be parsable in a few dozen lines even without any regexes, as @MStoykov's demo shows. We also don't need perfect UX, the current UX is so bad that almost anything will be a significant improvement :disappointed: Just take a look at the examples in https://github.com/grafana/k6/issues/961... So errors like `"foo" is an invalid threshold expression` are more than good enough, for now.

Again, we don't know how parsing of the future thresholds will look like. We might go with our own custom parser, and then a fancy fully-featured one will be great. But we might decide to adopt PromQL or some other widely-used format for them and use a ready library for parsing it. So, no need to over-engineer anything here, just to get rid of goja and have basic validation of the threshold syntax...

oleiade commented:
Thanks for the context clarification @na--. Makes sense. I'll strip down the implementation in accordance then.

However, something that the review up to this stage outlined is that there were different perceptions as to what the accepted format should be. At the moment, any expression that's valid JS is accepted (although I think we do not document it as such). Do we target the same level of compatibility to not break anything, or is it okay to support the absolute minimum even if that's effectively a breaking change in the user-facing interface?

Can we agree on a target specification for the expressions to make our(my) lives easier? Incorporating my understanding of the current documented state, and the outcome of the discussions that happened in this PR, I would expect this to not break anything for users:

```
assertion           -> aggregation_method whitespace* operator whitespace* float
aggregation_method  -> trend | rate | gauge | counter
counter             -> "count" | "sum" | "rate"
gauge               -> "last" | "min" | "max" | "value"
rate                -> "rate"
trend               -> "min" | "mean" | "avg" | "max" | percentile
percentile          -> "p(" float ")"
operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
float               -> digit+ (. digit+ | "e" -? digit+)?
digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
whitespace          -> space
space               -> " "
```

na-- commented:
We have to double check exactly what the cloud supports, but here are some things that I noticed:
- Gauge metrics don't have `last`, `min` and `max` in k6, [only `value`](https://github.com/grafana/k6/blob/703fd5723f51f5fb7e24e87361f920870d2e3564/stats/sink.go#L85)
- similarly, Counter [doesn't have `sum`](https://github.com/grafana/k6/blob/703fd5723f51f5fb7e24e87361f920870d2e3564/stats/sink.go#L60-L61)
- Trend [has `med`](https://github.com/grafana/k6/blob/703fd5723f51f5fb7e24e87361f920870d2e3564/stats/sink.go#L156)
- `aggregation_method` is actually just the metric type, it's not part of the parsing here, instead we need look up the metric type from the new `Registry` based on the metric name
- we probably don't need to support `E` notation for float numbers? 
- we can probably support more white-space options beyond just space, depending on the cloud :thinking: 

For the first 3 points, if the cloud already supports these "operators", we can probably also easily implement them, see https://github.com/grafana/k6/issues/1312

oleiade commented:
I've pushed a dumbed-down version of the parser as requested. 

- **regarding spaces**: the cloud drops any amount of whitespaces. As far as I can tell, it does not manage tabs, thus I haven't added them to the specification, nor support it in the parser.
- **regarding the exponential notation**: I've dropped any explicit support for it and rely on Go's [`strconv.ParseFloat`](https://pkg.go.dev/strconv#ParseFloat) method exclusively. My understanding is that `ParseFloat` does not support the exponential notation. 

It seems I copy-pasted an outdated definition indeed. Sorry for that. Here's the up to date one (incorporating your comments):

```
// assertion           -> aggregation_method  whitespace* operator whitespace* float
// aggregation_method  -> trend | rate | gauge | counter
// counter             -> "count" | "rate"
// gauge               -> "value"
// rate                -> "rate"
// trend               -> "avg" | "min" | "max" | "med" | percentile
// percentile          -> "p(" float ")"
// operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
// float               -> digit+ ("." digit+)?
// digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
// whitespace          -> " "
```

MStoykov commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,145 @@
+
+	// Otherwise, attempt to parse a percentile expression
+	if strings.HasPrefix(input, "p(") && strings.HasSuffix(input, ")") {
+		trimmed := strings.TrimSuffix(strings.TrimPrefix(input, "p("), ")")
```
Nit: I would've written this as `input[2:len(input)-1]` possibly replacing 2 and 1 with `len("p(")` 

Not particularly certain my shorter version is more readable as it's shorter, or your longer version is better with it being more specific/expressive  

I wonder if some helper function
```
func TrimSides(input, prefix, suffix string) (result string, didTrim bool) 
```
Will not be beneficial, I think I am doing this in other places as well.

But regardless, just mentioning it, it's not a requirement to be fixed.

oleiade replied:
My personal preference is to go with a more functional style and try to avoid "static" indices in code as much as I can, as they involve the overhead of thinking about what they mean.

I've added a private `trimDelimited(prefix, input, suffix string) string` helper function to hide the logic and allow a much clearer `trimDelimited("p(", input, ")")` call.

MStoykov commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,145 @@
+	// Otherwise, attempt to parse a percentile expression
+	if strings.HasPrefix(input, "p(") && strings.HasSuffix(input, ")") {
+		trimmed := strings.TrimSuffix(strings.TrimPrefix(input, "p("), ")")
+		_, err := strconv.ParseFloat(trimmed, 64)
```
This will parse floats in the form 3.1415927E+02 I am technically for us *parsing* this, but I remember there was some discussion, so I am mentioning it

oleiade replied:
You're right 👍🏻 If no blocking objections are made, I would keep it that way, as it relies on a standard library method. Otherwise, I would specifically look for the `e+` sequence and make the parser fail accordingly.

MStoykov reviewed (changes requested):
LGTM :clap: , but percentiles definitely need fixing ;)

I have left other nitpicks, but more tests with percentiles will be appreciated ;)

oleiade commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,145 @@
+// boils down to an expression of the form: `p(float64)`, but will return
+// it verbatim, as a string.
+func parseThresholdAggregationMethod(input string) (string, error) {
+	methods := []string{
```
Indeed! I've defined the tokens as constant strings, and the methods as global array `[7]string` instead 👍🏻

oleiade commented on stats/thresholds_parser.go:
```diff
@@ -0,0 +1,145 @@
+// side value will be made: they will be returned as is, only trimmed from
+// their spaces.
+func scanThresholdExpression(input string) (string, string, string, error) {
+	operators := []string{
```
Indeed! I've defined the operator tokens as constant strings, the operators as a global `[7]string` array, and documented that the operator order matters 👍🏻

MStoykov commented on stats/thresholds.go:
```diff
@@ -138,60 +154,101 @@ func NewThresholds(sources []string) (Thresholds, error) {
+		ts.sinked["max"] = sinkImpl.Max
+		ts.sinked["avg"] = sinkImpl.Avg
+		ts.sinked["med"] = sinkImpl.Med
+		ts.sinked["p(90)"] = sinkImpl.P(0.90)
+		ts.sinked["p(95)"] = sinkImpl.P(0.95)
```
This shouldn't be needed if the below part will add the *one* percentile it needs

This seems like a good place for some optimization (not in this PR, this was the case before as well), but here we do generate a whole map with a bunch of values. But in most cases only 1 of those will be used at all.

This is *low prio* though IMO, only the percentiles calculations are "heavy" and as long as you drop this 2 precalculated which might not be needed, the below code will calculate just the ones it needs.

Also given that in this case it will calculate the percentile ones even if they are used in multiple thresholds, it might even not be beneficial to JIT calculate them if we also don't cache them :thinking: 

But I digress – just drop those two lines 

```suggestion
```

oleiade replied:
Yeah, I had doubts about this too, but put them back as they are "default" percentiles, and wanted to make no assumptions to whether they would be present or not. Glad you clarified it, I'll drop them back indeed! 👍🏻

MStoykov commented on stats/thresholds_parser_test.go:
```diff
@@ -0,0 +1,144 @@
+	"testing"
+)
+
+func Test_parseThresholdExpression(t *testing.T) {
```
Is there a particular reason you decided to go with `Test_xxx` instead of what seems to be the standard `TestXxx` naming that we've been using through k6? 

I am not really particularly for one or the other, but I prefer if there is one that we stick to instead of it being different everywhere ;)

oleiade replied:
Good catch!

We had this discussion internally a while ago, and really decided that test names shouldn't have an underscore separator, indeed.  I think these tests predate that discussion (if not, I'll blame VSCode's snippets instead).

MStoykov commented on stats/thresholds_parser_test.go:
```diff
@@ -0,0 +1,144 @@
+		wantCondition *thresholdExpression
+		wantErr       bool
+	}{
+		{"unknown expression's operator fails", "count!20", nil, true},
+		{"unknown expression's method fails", "foo>20", nil, true},
+		{"non numerical expression's value fails", "count>abc", nil, true},
+		{"valid threshold expression syntax", "count>20", &thresholdExpression{AggregationMethod: "count", Operator: ">", Value: 20}, false},
```
nit: I in general do not like "keyless" struct literals, both because they are harder to read and because adding one filed means you need to change all of them, even if most are fine with their zero value.

I do think that in this case (especially with some single digit tests) it's fine and maybe even better for readability. If any other reviewer also wants them to be with keys, perhaps we should change it though :shrug: 

The other cases in this file are getting more and more in the territory of me wanting them to not be keyless.

oleiade replied:
I was unsure about the established best practice. Pretty sure I've seen both styles throughout the codebase, but my memory could also be failing me. In general, and in the spirit of "explicit is better than implicit", I agree with you.

I'll add the struct keys 😸

MStoykov commented on stats/thresholds_parser_test.go:
```diff
@@ -0,0 +1,144 @@
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+
+			gotCondition, err := parseThresholdExpression(testCase.input)
+			if (err != nil) != testCase.wantErr {
+				t.Errorf("parse() error = %v, wantErr %v", err, testCase.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(gotCondition, testCase.wantCondition) {
+				t.Errorf("parse() = %v, want %v", gotCondition, testCase.wantCondition)
+			}
```
Is there any reason why you are not using [testify](https://pkg.go.dev/github.com/stretchr/testify) we are after all using it through the codebase?

oleiade replied:
Once again, I'll blame VSCode. The Go extension's `generate unit test for function` command does use the standard package only; and that's the one I use to write all the boilerplate for me. I'll modify the code to use `testify` instead, and find out if I can customize the snippet command to avoid that in the future 👍🏻

MStoykov commented on stats/thresholds_test.go:
```diff
@@ -25,99 +25,250 @@ import (
+			false,
+		},
+		{
+			"valid expression over non-existing sink",
+			fields{"rate>0.01", false, false, types.NullDurationFrom(2 * time.Second), &thresholdExpression{tokenRate, tokenGreater, 0.01}},
+			args{map[string]float64{"med": 27.2}},
+			false,
+			true,
```
I definitely want to know what the fifth element here is without having to go up to the definition of the type.

oleiade replied:
Agreed 👍🏻

MStoykov reviewed (changes requested):
Looks much better IMO :clap:. I

 have some comments around testing, some more nitpicky others less nitpicky but it seems like most of them aren't a single case, and I opted for not pointing out all the cases before we discuss them.

oleiade commented:
@MStoykov Thanks a lot for your review. I have addressed your comments. Let me know if you have other comments, recommendations or questions 👍🏻 

Ping @na-- @yorugac The PR has been rewritten for the most part, and is now in a reviewable state again. Please take a look 👀 whenever you've got a moment 😸

MStoykov commented on stats/thresholds_test.go:
```diff
@@ -25,99 +25,274 @@ import (
+	}{
+		{
+			name:    "valid expression using the > operator over passing threshold",
+			fields:  fields{"rate>0.01", false, false, types.NullDurationFrom(2 * time.Second), &thresholdExpression{tokenRate, tokenGreater, 0.01}},
```
Here `fields` has a bunch of ... fields (:stuck_out_tongue: ) three of which are the same in every case. They also coincide with the fields of Threshold.

From the looks of it - the one that is actually used is also `parsed` which IMO can just be part of the top test case struct. 
Also see https://github.com/grafana/k6/pull/2251/files#r756909900 although from my understandign because this is `runNoTaint` that isn't really interesting parameter. Still (maybe an additional test) to check that it doesn't matter will bea good idea :shrug: .

oleiade replied:
For fields, VSCode snippets are at it again, you're right they're quite redundant. Going to address that indeed🙏🏻

MStoykov commented on stats/thresholds_test.go:
```diff
@@ -25,99 +25,274 @@ import (
+		{
+			name:    "valid expression using the > operator over passing threshold",
+			fields:  fields{"rate>0.01", false, false, types.NullDurationFrom(2 * time.Second), &thresholdExpression{tokenRate, tokenGreater, 0.01}},
+			args:    args{map[string]float64{"rate": 1}},
```
Seems to me like you can just make `args` `map[string]float64` and skip some types and wrapping in the process

oleiade replied:
Agreed 👍🏻

MStoykov commented on stats/thresholds.go:
```diff
@@ -138,60 +154,99 @@ func NewThresholds(sources []string) (Thresholds, error) {
-	succ := true
-	for i, th := range ts.Thresholds {
-		b, err := th.run()
+func (ts *Thresholds) runAll(duration time.Duration) (bool, error) {
```
I would argue `duration` doesn't sound correct as this is the ["current test run duration"](https://github.com/grafana/k6/blob/c93adb3732ddb182ade693191bfe56ad0cfc1e96/core/engine.go#L449-L459). You might call it the duration so far, but I would argue `timeInTest` will be slightly better as "current" might be a bit misleading given that it's (at this point) somewhere in the past ;)

oleiade replied:
Thanks for the recommendation 👍🏻 I've renamed the parameter to `timeSpentInTest`, and it does make the code itself much clearer indeed 😸

oleiade commented on stats/thresholds_test.go:
```diff
@@ -22,102 +22,284 @@ package stats
+		Source           string
+		LastFailed       bool
+		AbortOnFail      bool
+		AbortGracePeriod types.NullDuration
```
Indeed, and as a result of @MStoykov's recommendation, I added a specific test for it 😸

yorugac reviewed (changes requested):
I think overall it's all clear and good :tada: My request changes are very minor but there are also a couple of more involved questions too, please see.

Also, out of curiosity: I ran benchmarks added in here and they perform nicely, but is there a way to compare it easily to the old version?

MStoykov commented on stats/thresholds_parser_test.go:
```diff
@@ -0,0 +1,333 @@
+			name:         "expression with unknown operator fails",
+			input:        "foo!bar",
+			wantMethod:   "",
+			wantOperator: "", wantValue: "",
```
Hm, I would argue they can just be left ... out and they will be empty ;)

oleiade replied:
I prefer to keep them for it to be explicit and document that they won't be set in this context 😸

oleiade commented on stats/thresholds.go:
```diff
@@ -138,60 +154,99 @@ func NewThresholds(sources []string) (Thresholds, error) {
+				continue
+			}
+
+			// Percentile value has already been parsed as a float64, and should be valid.
+			// The error check is only here to protect our future selves.
+			percentile, err := strconv.ParseFloat(trimDelimited("p(", threshold.parsed.AggregationMethod, ")"), 64)
+			if err != nil {
+				return false, fmt.Errorf(
+					"unable to parse percentile value as 64-bit floating point number; "+
+						"reason: %w", err,
+				)
+			}
```
The idea was to make a tradeoff. Considering that percentile are a "special-case", as they hold an actual value, and are the only ones doing so, I preferred parsing them as a "validation that the expression makes sense", and extract their value only later "just in time", as they're actually used. That way, I avoided having a dedicated nullable value in the `thresholdExpression`.

I can also envision a version of this where we introduce a `tokenPercentile`, and a `null.Float` (if it exists) in the `thresholdExpression`. We would then in the context of this version iterate over the thresholds, and if the aggregation method is a percentile, then we verify the value is actually set and add it to the sinks. Would you prefer it that way? 😸 

I truly have no preference. I can see on one side, we pre-mash the work, while the other avoids adding abstraction. Let  me know 🐻

yorugac replied:
Hm, looking at it with fresh eyes, yes, you're right, it can be seen either way... I think my first "instinct" is to minimize number of possible errors and parsing which would be with an additional field in struct. It also reduces number of lines :smile: But depending on how thresholds evolve in the future, that field might become obsolete / to be moved, etc.
So yes, I have a slight preference towards the `tokenPercentile` field but not a very strong one, in case you change your mind and would like to keep the current version :slightly_smiling_face:

oleiade replied:
I've modified the implemented design accordingly 👍🏻. The `thresholdExpression` type now has an `AggregationValue` field that's a `null.Float`, and the parser defines a `tokenPercentile` constant too. Let me know if it looks good on your side 😸

yorugac replied:
Yes, I think it "reads" well and easy! Thank you :slightly_smiling_face:

oleiade commented on stats/thresholds.go:
```diff
@@ -138,60 +154,99 @@ func NewThresholds(sources []string) (Thresholds, error) {
+	ts.sinked = make(map[string]float64)
+
+	// FIXME: Remove this comment as soon as the stats.Sink does not expose Format anymore,
+	// see: https://github.com/grafana/k6/pull/2251#discussion_r769807203
```
I have opened a dedicated issue: #2320, and updated the comment accordingly 👍🏻

oleiade commented:
Note that I will make sure to squash all "Update", and "fixup!" commits before merging 👍🏻

na-- commented on stats/thresholds.go:
```diff
@@ -73,23 +54,57 @@ func newThreshold(src string, newThreshold *goja.Runtime, abortOnFail bool, grac
+	// expression
+	lhs, ok := sinks[t.parsed.AggregationMethod]
+	if !ok {
+		return false, fmt.Errorf("unable to apply threshold %s over metrics; reason: "+
```
The next step here, in another PR, would be to tie the threshold validation with the metrics in our metrics registry. This validation is awesome, much better than what existed (or rather, what didn't exist :sweat_smile:) before, but we can do it right after the initial execution of the init context. Right now, this script:
```js
import { sleep } from 'k6';
import { Counter } from 'k6/metrics';
import exec from 'k6/execution';


const myCounter = new Counter('my_counter');

export const options = {
    vus: 5,
    iterations: 30,
    thresholds: {
        my_counter: ['p(95)<200'], // error, Counters do not have p()
    },
};


export default function () {
    console.log(`[t=${(new Date()) - exec.scenario.startTime}ms] VU{${exec.vu.idInTest}} ran scenario iteration ${exec.scenario.iterationInTest}`);
    myCounter.add(1);
    sleep(1);
}
```

will emit something like this:
```
INFO[0000] [t=0ms] VU{3} ran scenario iteration 0        source=console
INFO[0000] [t=0ms] VU{5} ran scenario iteration 2        source=console
INFO[0000] [t=0ms] VU{2} ran scenario iteration 1        source=console
INFO[0000] [t=0ms] VU{1} ran scenario iteration 3        source=console
INFO[0000] [t=0ms] VU{4} ran scenario iteration 4        source=console
INFO[0001] [t=1001ms] VU{2} ran scenario iteration 7     source=console
INFO[0001] [t=1001ms] VU{4} ran scenario iteration 5     source=console
INFO[0001] [t=1001ms] VU{3} ran scenario iteration 9     source=console
INFO[0001] [t=1001ms] VU{5} ran scenario iteration 6     source=console
INFO[0001] [t=1001ms] VU{1} ran scenario iteration 8     source=console
ERRO[0002] Threshold error                               component=engine error="threshold 0 run error: unable to apply threshold p(95)<200 over metrics; reason: no metric supporting the p(95) aggregation method found" m=my_counter
```

but right now there is nothing preventing us from validating the thresholds _before_ we actually start running the script, `registry` should have all of the metrics (custom or built-in) and their types right after the last line here: https://github.com/grafana/k6/blob/23575eb63c948afc34dbe58a5e24d803305b4803/cmd/run.go#L113-L115

na-- replied:
Fixing https://github.com/grafana/k6/issues/1346 can probably be done at the same time as :arrow_up: , but I'm not finding an already open issue for it. @oleiade, do you want to create one or should I?

na-- replied:
And, thinking about it, https://github.com/grafana/k6/issues/2320 will also be easier when we don't have to do any validation in `Thresholds.Run()` :thinking: :tada: Each threshold can be "pre-compiled" by just turning it into a Go lambda or something like that. Once we do the validation once after the init context execution, that a threshold impression is valid for the metric type it is defined for, we shouldn't need any more validation afterwards :tada:

oleiade replied:
Very good points indeed! I've created #2330 to keep track of it 🎉

na-- reviewed (approved):
LGTM, so much nicer than the JS hack :heart: :sweat_smile: it also unlocks us to implement so many other fixes and improvements like https://github.com/grafana/k6/pull/2251/files#r782875298, https://github.com/grafana/k6/issues/1346 and threshold improvements in general :tada:

oleiade commented:
We experienced issue with this feature in the larger, internal, context of K6, and decided to revert it, and delay its release to `v0.37.0`. See #1443 for more details.
